File: leet/0783-minimum-distance-between-BST-nodes/minimum-distance-between-BST-nodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDiffInBST(root *TreeNode) int {
    ints := inorderTraversal(root)
    
    if (len(ints) < 2) {return 0 }
    
    rs := 0
    last := 0
    for i, num := range ints {
        diff := num - last
        last = num
        if(i == 0) { continue }
        if(i == 1) { rs = diff; continue}
        if (rs > diff) {rs = diff}
    }
    
    return rs
}

func inorderTraversal(root *TreeNode) []int {
    if(root == nil) {return []int{}}
    
    rs := make([]int, 0)
    rs = append(rs, inorderTraversal(root.Left)...)
    rs = append(rs, root.Val)
    rs = append(rs, inorderTraversal(root.Right)...)
    
    return rs
}

File: leet/0783-minimum-distance-between-BST-nodes/minimum-distance-between-BST-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function(root) {
    let getMax = function(node){
        if (node.right) return getMax(node.right)
        return node
    }
    
    let getMin = function(node){
        if (node.left) return getMin(node.left)
        return node
    }
    let ret = 999999
    if (root.left) ret = Math.min(ret, root.val - getMax(root.left).val)
    if (root.right) ret = Math.min(ret, getMin(root.right).val - root.val)
    
    if(root.left) ret = Math.min(ret, minDiffInBST(root.left))
    if(root.right) ret = Math.min(ret, minDiffInBST(root.right))
    
    return ret
};
